lsp/nls/src/cache.rs

```diff
@@ -48,11 +49,59 @@ impl CacheExt for Cache {
         if let Ok(CacheOp::Done((ids, errors))) = self.resolve_imports(file_id) {
             import_errors = errors;
             for id in ids {
-                self.typecheck_with_analysis(id, initial_ctxt, initial_env, lin_cache)
-                    .unwrap();
+                // Ignore the results, and check for errors after resolution
+                let _ = self.typecheck_with_analysis(id, initial_ctxt, initial_env, lin_cache);
             }
         }
 
+        // After self.parse(), the cache must be populated
+        let CachedTerm {
+            term,
+            state,
+            parse_errs,
+        } = self.terms_mut().remove(&file_id).unwrap();
+
+        // We do this to get a list of the imports that have been resolved
+        // but don't typecheck, and then, we give an appriopiate diagnostics
+        let mut errors = Vec::new();
+        let term = term.traverse::<_, _, ()>(
```

Comment:
I don't think traversing the whole term again _unconditionally_ (even if no import failed) is reasonable. Some terms might be huge (@vkleen's auto-generated contracts).

If you need this information, would that be possible to build it up during linearization? Keeping e.g. a list of tuples `(location, import_id)`, or even a map  `HashMap<FileId, Vec<TermPos>>`. I won't include the imports that are in a file which failed to parse or typecheck, but I think it's fine for now: at least, a well-typed file will show if there is an issue with its imports.

Secondly, we might want to know if the previous `for` succeeded, in case we can avoid to do this search in the first place.

Reply:
Because it seems harder to do. If you build the import during linearization, but the current file fails to typecheck, you might not have visited all the terms and the import node.

Also, the need is maybe less clear cut (at least to me): if the current file doesn't even typecheck, the user has already some local red squiggles to work on anyway. Then, once the current file typechecks, we can show imports that has failed, in a second time.

So, put differently, I believe ignoring the import statements in a file which doesn't typecheck seems easier to do, and the consequence doesn't seem too bad.